modules/c++/io/include/io/BufferViewStream.h

```diff
@@ -0,0 +1,191 @@
+     * \param whence Location to seek from
+     * \return new position
+     */
+    virtual sys::Off_T seek(sys::Off_T offset, Whence whence);
+
+    /*
+     * \return The available bytes to read from the stream
+     */
+    virtual sys::Off_T available()
+    {
+        return mBufferView.size - mPosition;
+    }
+
+    using OutputStream::write;
+    using InputStream::streamTo;
+
+    /*
+     * Writes the bytes in data to the stream.
+     * \param buffer The data to write to the stream
+     * \param size The number of bytes to write to the stream
+     */
+    virtual void write(const void* buffer, size_t size);
+
+    /*!
+     * Get a pointer to the internal buffer.
+     * This pointer should not be treated as valid
+     * after a call to the seek, write, or reset methods
+     * \return pointer to the internal buffer
+     */
+    sys::ubyte* get()
+    {
+        return mBufferView.data;
+    }
+
+protected:
+    /*!
+     * Read up to len bytes of data from this buffer into an array
+     * update the mark
+     * \param buffer Buffer to read into
+     * \param len The length to read
+     * \return  The number of bytes read
+     */
+    virtual sys::SSize_T readImpl(void* buffer, size_t len);
+
+
+private:
+    mem::BufferView<T>& mBufferView;
+    sys::Off_T mPosition;
+};
+
+template <typename T>
+sys::Off_T BufferViewStream<T>::seek(sys::Off_T offset, Whence whence)
+{
+    // Let's not change anything until we know it will be valid
+    sys::Off_T newPos = mPosition;
+    switch (whence)
+    {
+        case START:
+            newPos = offset;
+            break;
+        case END:
+            if (offset > static_cast<sys::Off_T>(mBufferView.size))
+            {
+                newPos = 0;
+            }
+            else
+            {
+                newPos = mBufferView.size - offset;
+            }
+            break;
+        default:
+            newPos += offset;
+            break;
+    }
+
+    if (newPos > static_cast<sys::Off_T>(mBufferView.size) || newPos < 0)
+    {
+        throw except::Exception(Ctxt("Attempted to seek beyond end of stream"));
+    }
+    mPosition = newPos;
+    return tell();
+}
+
+template <typename T>
+void BufferViewStream<T>::write(const void* buffer, size_t size)
+{
+    const sys::Size_T newPos = mPosition + size;
+    if (newPos > mBufferView.size)
+    {
+        std::ostringstream msg;
+        msg << "Write of size " << size << " runs out of bounds.";
+        throw except::Exception(Ctxt(msg.str()));
+    }
+
+    const T* const bufferPtr = static_cast<const T*>(buffer);
+    std::copy(bufferPtr, bufferPtr + size, mBufferView.data + mPosition);
+    mPosition = newPos;
+}
+
+template <typename T>
+sys::SSize_T BufferViewStream<T>::readImpl(void* buffer, size_t len)
+{
+    if (available() < static_cast<sys::Off_T>(len))
+    {
+        len = available();
+    }
+    if (len == 0)
+    {
+        return 0;
+    }
+
+    ::memcpy(buffer, mBufferView.data + mPosition, len);
```

Comment:
Be consistent - use either `std::copy()` or `::memcpy()` throughout

Reply:
I took the `std::copy` from `io::ByteStream`.... are we allowed to use C++11 in here now, or should I change that to `memcpy` while I'm in here?

Reply:
Still avoiding C++11 on the stuff on GitHub to support older compilers but `std::copy()` has been around since forever:

https://en.cppreference.com/w/cpp/algorithm/copy